lib/elixir/lib/module/types/descr.ex

```diff
@@ -84,14 +83,27 @@ defmodule Module.Types.Descr do
   @term_or_optional %{bitmap: @bit_top ||| @bit_optional, atom: @atom_top, map: @map_top}
 
   def not_set(), do: @not_set
-  def if_set(type), do: Map.update(type, :bitmap, @bit_optional, &(&1 ||| @bit_optional))
   defp term_or_optional(), do: @term_or_optional
 
+  def if_set(:term), do: term_or_optional()
+  def if_set(type), do: Map.update(type, :bitmap, @bit_optional, &(&1 ||| @bit_optional))
+
+  defguardp is_optional(map)
+            when is_map(map) and
+                   ((is_map_key(map, :bitmap) and (map.bitmap &&& @bit_optional) != 0) or
+                      (is_map_key(map, :dynamic) and is_map(map.dynamic) and
+                         is_map_key(map.dynamic, :bitmap) and
+                         (map.dynamic.bitmap &&& @bit_optional) != 0))
+
+  defguardp is_optional_static(map)
+            when is_map(map) and is_map_key(map, :bitmap) and (map.bitmap &&& @bit_optional) != 0
```

Comment:
```suggestion
  defguardp is_optional_static(map)
            when is_map(map) and is_map_key(map, :bitmap) and (map.bitmap &&& @bit_optional) != 0

  defguardp is_optional(map)
            when is_optional_static(map) or
                   (is_map(map) and is_map_key(map, :dynamic) and is_optional_static(map.dynamic))
```
possibly easier to read

Reply:
I will check later if the Erlang compiler optimizes the duplicate checks efficiently and merge your proposal. :)